### PR TITLE
[SPARK-11859][Mesos] SparkContext accepts invalid Master URLs in the form zk://host:port for a multi-master Mesos cluster using ZooKeeper

### PR DESCRIPTION
**Pull request by toddwan**
- According to below doc and validation logic in [SparkSubmit.scala](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L231), master URL for a mesos cluster should always start with `mesos://` 

http://spark.apache.org/docs/latest/running-on-mesos.html
`The Master URLs for Mesos are in the form mesos://host:5050 for a single-master Mesos cluster, or mesos://zk://host:2181 for a multi-master Mesos cluster using ZooKeeper.`
- However, [SparkContext.scala](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkContext.scala#L2749) fails the validation and can receive master URL in the form `zk://host:port`
- For the master URLs in the form `zk:host:port`, the valid form should be `mesos://zk://host:port`
- This PR restrict the validation in `SparkContext.scala`, and now only mesos master URLs prefixed with `mesos://` can be accepted.
- This PR also updated corresponding unit test.

### CODE REVIEW

**Comment by dragos** [comment]
@andrewor14  I wonder if we shouldn't first warn about this, and defer the actual failure until 2.0. There might be people relying on this loophole. If I understand correctly, people could connect using `zk://` urls if they didn't go through spark-submit (spark-shell or hard coded).

**Comment by srowen** [comment]
LGTM, though I tend to agree there's a little risk here in making something that shouldn't work actually not work.

**Comment by srowen** [comment]
@toddwan what do you think about writing a separate code path to handle the incorrect `zk://...` syntax? It could use the same "case", if it's a regex that captures the whole string. And then embed a warning if the actual arg starts with `zk://`

**Comment by tnachen** [comment]
+1 on having a fall back with a warning message as well.

**Comment by toddwan** [comment]
A new commit has been appended to the PR. Now a warning message is printed first if the given mesos Master URL starts with `zk://`, then a correct mesos Master URL is constructed by prefixing `mesos://` to the current one, and pass the new URL to `SparkContext.createTaskScheduler()` again.

**Comment by srowen** [comment]
Yes, that's pretty nice and clean.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
LGTM, I'll merge this once tests pass.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2727,6 +2726,11 @@ object SparkContext extends Logging {
         scheduler.initialize(backend)
         (backend, scheduler)
 
+      case zkUrl if zkUrl.startsWith("zk://") =>
+        logWarning("Master URL for a multi-master Mesos cluster managed by ZooKeeper should be " +
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Nitpicking, but I would be more explicit: "This will stop working in Spark 2.0".

**Comment by dragos** [comment]
other than that, LGTM

**Comment by toddwan** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by toddwan** [comment]
@andrewor14 I am not sure about the cause of the unit test failure, but the failed test reported in the latest test build can pass on my machine.

**Comment by srowen** [comment]
LGTM, will merge soon.

**Comment by dragos** [comment]
LGTM

**Comment by srowen** [comment]
@toddwan let me know your JIRA handle and I'll credit you. Merged to master/1.6

**Comment by tawan0109** [comment]
@srowen https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tawan thanks.

**Pull request closed without merging**